### ISSUE TITLE
Summarise thoughts since last meeting

### ISSUE THREAD

**Issue by CJcool06** [ChrisFuscoMasters/TransformerImpl#13]
I should summarise my thoughts on Transformers and also write comments on the papers that I've read.

**Comment by CJcool06** [comment]
**Transformer Thoughts**

- Transformers have no/little inductive biases. CNN has translation invariance, RNN has temporal invariance, and both use locality.

- The attention block maps input features to queries, keys, and values. Dot-product attention decides how much weight should be applied to the values for each feature. Conceptually, this is memory retrieval with an extra step of allowing features to share values. The sharing (weight) for a feature is decided by the similarity of its query with every other feature's key (and its own). At least, this is how I currently understand it.

- It's because of these two reasons that Transformers are data hungry: **1)** Required to learn inductive biases that come out-of-the-box with other architectures **2)** Required to build up a memory bank of queries, keys, and values. Poor performance may correlate with sparse memory.

- The attention mechanism is very powerful and can/has be successfully applied across sub-domains. It's less clear to me if the same could be said about the entire Transformer architecture.

**Comment by CJcool06** [comment]
**Papers**

- Deep Residual Learning for Image Recognition (Great)

Reformulates NN layers as learning residual functions with reference to the layer inputs, instead of learning unreferenced functions. The intuition is that if the layers of a network produces _x_ training error, then adding N layers on top of that model should produce at most _x_ training error. Conceptually, these added layers can _identity map_ and produce the same output. To ensure that this is the case, they introduce a _deep residual learning_ framework by adding shortcut connections. This was a seminal paper in the field.

- Adam: A Method for Stochastic Optimization (Amazing)

An algorithm for first-order gradient-based optimisation. The method computes individual adaptive learning rates for different parameters from estimates of first and second moments of the gradients. Some of Adam's advantages are that the magnitudes of parameter updates are invariant to rescaling of the gradient, its step-sizes are approximately bounded by the step-size hyperparameter, it does not require a stationary objective, it works with sparse gradients, and it naturally performs a form of step-size annealing. Conceptually, if we are unsure of the direction of the optimal solution, take smaller steps.

- An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale (Good)

The authors present a pure-transformer method for image classification, attempting to remove the reliance of CNNs. The method is to cut an image up into N same-size patches, flatten the patches, apply patch and positional encodings, and attach an MLP head. They also attach an extra learnable "classification token" to the input sequence that is used by the MLP head. Their method outperforms CNN methods on large datasets not small/mid-scale datasets.

- SegViT: Semantic Segmentation with Plain Vision Transformers (Ok)

ViT is used as a backbone to generate image features. Their architecture is a 3-layer modified decoder block that uses the features from each ViT decoder layer as key and value tokens, and class encodings from the previous layer as queries. Each decoder block outputs a mask from their modified attention sub-block and also new class encodings. The encodings from the final layer are fed through a linear transformation and apply softmax to get the class probabilities.

- 3D Point Positional Encoding for Multi-Camera 3D Object Detection Transformers (Ok)

They use the DETR architecture and method. They devise a new method for encoding 3D point positions to the inputs. Their method performs better than other DETR-based methods for object detection.

- BEVFusion: A Simple and Robust LiDAR-Camera Fusion Framework (Meh)

The proposed method fuses multiple camera feeds and lidar feed in a disentangled manner, thus performing better in scenarios where the lidar has malfunctioned. Not too much positive to say about this one, it was very ad-hoc and they didn't demonstrate the method's effectiveness properly (used different training schemes for comparisons, thus results were not trustworthy).

- A Survey of Transformers (Great)

This survey closely looks at the Transformer architecture and discusses current areas of research.

- Transformers in Vision: A Survey (Great)

This survey looks at current research in applying the Transformer architecture for computer vision problems.

- Lift, Splat, Shoot

The core idea behind the approach is to "lift" each image individually into a frustrum of features for each camera, then "splat" all frustrums into a rasterised birds-eye-view grid. Their method uses a learned depth distribution, which was interesting. 

- Low-Rank and Locality Constrained Self-Attention for Sequence Modelling

They propose the use of two types of attention inside an attention block - band attention and low-rank attention. Their formulation of their self-attention changes using matrix decomposition was nice as they showed why they did what they did. Although the maths was nice, their conclusions were not often backed up. Also, their final model architecture and evaluation schema did not inspire confidence in their hypothesis.

Comments on more recently read papers can be found in #12 and #9.
